File: Excel/1000-1099/1017/Challenge 1017.py
import pandas as pd
import re

path = "1000-1099/1017/1017 Typographical Changes.xlsx"
input = pd.read_excel(path, usecols="A", nrows=19)
dictionary = pd.read_excel(path, usecols="C:D", nrows=36)
test = pd.read_excel(path, usecols="F", nrows=19)

dict_map = dict(zip(dictionary["Word"].str.lower(), dictionary["WordType"]))


def format_word(word, force_cap=False):
    punct = re.search(r"[,:.!?]+$", word)
    punct = punct.group(0) if punct else ""
    raw = re.sub(r"[,:.!?]+$", "", word)
    is_bracketed = re.fullmatch(r"\[.+\]", raw) is not None
    clean = raw.strip("[]").lower()
    if is_bracketed:
        out = clean.upper()
    elif dict_map.get(clean) == "Acronym":
        out = clean.upper()
    elif dict_map.get(clean) == "Minor" and not force_cap:
        out = clean
    else:
        out = clean.capitalize()

    return f"[{out}]{punct}" if is_bracketed else f"{out}{punct}"


def format_title(text):
    words = text.split()
    force_cap = [i == 0 or words[i - 1].endswith(":") for i in range(len(words))]
    return " ".join(format_word(word, cap) for word, cap in zip(words, force_cap))


result = input.assign(Output=input["Data"].map(format_title))

print(result["Output"].equals(test["Answer Expected"]))
# inconsistent usage of bracket rule

File: Excel/1000-1099/1017/Challenge 1017.R
library(tidyverse)
library(readxl)

path <- "1000-1099/1017/1017 Typographical Changes.xlsx"
input <- read_excel(path, range = "A1:A20")
dictionary <- read_excel(path, range = "C1:D36")
test <- read_excel(path, range = "F1:F20")

minor <- dictionary %>%
  filter(WordType == "Minor") %>%
  pull(Word) %>%
  str_to_lower()

acronyms <- dictionary %>%
  filter(WordType == "Acronym") %>%
  pull(Word) %>%
  str_to_lower()

fix_word <- \(x, force_cap = FALSE) {
  punct <- str_extract(x, "[,:.!?]+$") %>% replace_na("")
  word <- str_remove(x, "[,:.!?]+$")

  bracketed <- str_detect(word, "^\\[.+\\]$")
  core <- word %>% str_remove("^\\[") %>% str_remove("\\]$")
  key <- str_to_lower(core)

  out <- case_when(
    bracketed ~ str_to_upper(core),
    key %in% acronyms ~ str_to_upper(core),
    key %in% minor & !force_cap ~ str_to_lower(core),
    TRUE ~ str_to_title(str_to_lower(core))
  )

  str_c(if_else(bracketed, str_c("[", out, "]"), out), punct)
}

title_fix <- \(x) {
  words <- str_split_1(x, "\\s+")

  map2_chr(
    words,
    seq_along(words) == 1 | lag(str_detect(words, ":$"), default = FALSE),
    fix_word
  ) %>%
    str_c(collapse = " ")
}

result <- input %>%
  mutate(Output = map_chr(Data, title_fix))

all.equal(result$Output, test$`Answer Expected`)
# inconsistent usage of bracket rule
